core/node/node_framework/src/implementations/layers/postgres.rs

```diff
@@ -111,62 +112,60 @@ impl From<DatabaseInfo> for Health {
 }
 
 #[derive(Debug)]
-pub struct DatabaseHealthTask {
+struct DatabaseHealthCheck {
     polling_interval: Duration,
-    connection_pool: ConnectionPool<Core>,
-    updater: HealthUpdater,
+    pool: ConnectionPool<Core>,
+    cached: RwLock<Option<(DatabaseInfo, Instant)>>,
 }
 
-impl DatabaseHealthTask {
-    fn new(connection_pool: ConnectionPool<Core>) -> Self {
-        Self {
-            polling_interval: TASK_EXECUTION_INTERVAL,
-            connection_pool,
-            updater: ReactiveHealthCheck::new("database").1,
-        }
+impl DatabaseHealthCheck {
+    async fn update(&self) -> anyhow::Result<DatabaseInfo> {
+        let mut conn = self.pool.connection_tagged("postgres_healthcheck").await?;
+        let last_migration = conn.system_dal().get_last_migration().await?;
+        Ok(DatabaseInfo { last_migration })
     }
 
-    async fn run(self, mut stop_receiver: watch::Receiver<bool>) -> anyhow::Result<()>
-    where
-        Self: Sized,
-    {
-        let timeout = self.polling_interval;
-        let mut conn = self
-            .connection_pool
-            .connection_tagged("postgres_healthcheck")
-            .await?;
-
-        tracing::info!("Starting database healthcheck with frequency: {timeout:?}",);
-
-        while !*stop_receiver.borrow_and_update() {
-            let last_migration = conn.system_dal().get_last_migration().await?;
-            self.updater.update(DatabaseInfo { last_migration }.into());
-
-            // Error here corresponds to a timeout w/o `stop_receiver` changed; we're OK with this.
-            tokio::time::timeout(timeout, stop_receiver.changed())
-                .await
-                .ok();
+    fn validate_cache(&self, cache: Option<&(DatabaseInfo, Instant)>) -> Option<DatabaseInfo> {
+        let now = Instant::now();
+        if let Some((cached, cached_at)) = cache {
+            let elapsed = now
+                .checked_duration_since(*cached_at)
+                .unwrap_or(Duration::ZERO);
+            (elapsed <= self.polling_interval).then(|| cached.clone())
+        } else {
+            None
         }
-        tracing::info!("Stop signal received; database healthcheck is shut down");
-        Ok(())
-    }
-
-    pub fn health_check(&self) -> ReactiveHealthCheck {
-        self.updater.subscribe()
     }
 }
 
-#[async_trait::async_trait]
-impl Task for DatabaseHealthTask {
-    fn kind(&self) -> TaskKind {
-        TaskKind::UnconstrainedTask
+#[async_trait]
+impl CheckHealth for DatabaseHealthCheck {
+    fn name(&self) -> &'static str {
+        "database"
     }
 
-    fn id(&self) -> TaskId {
-        "database_health".into()
-    }
+    // If the DB malfunctions, this method would time out, which would lead to the health check marked as failed.
```

Comment:
Am I reading this right that this only checks the replica pool? 

In theory, there shouldn't be such intermittent DB issues - particularly for the master DB. I think the node would crash if this happens, so `NotReady` is justified. As for replica - perhaps `Affected` is more appropriate. 

Having said that, this may be an overcomplication. How is this health check going to be used exactly? Maybe we should base the logic on the expected behaviour.

Reply:
AFAIU, the primary use case for the check is getting health details (i.e., information about the latest applied migration); measuring connectivity to the DB is a side effect (IMO, quite useful). It's always used for replica pool, at least for now. Let's merge the PR as is, and if liveness becomes an issue, rework the check.